doc/design/scope.md

```diff
@@ -0,0 +1,58 @@
+# Scope
+
+### Define
+
+Scope is a context to manage Variables. It mainly contains a map from Variable name to Variable. Net will get and update variable throw scope.
+
+```cpp
+class Variable;
+using VariablePtr = std::shared_ptr<Variable>;
+
+class Scope final {
+public:
+  Scope();
+  Scope(const std::shared_ptr<Scope>& parent);
+
+  //! Get Variable in this scope.
+  //! @return nullptr if no such variable.
+  const VariablePtr& getVar(const std::string& name) const;
```

Comment:
Here the return type must be `Variable*` if we use `unique_ptr` other than `shared_ptr` in `vars_`.

Also, following Google code style, this method should be named `GetVar`.